### PR TITLE
unsubscribe and unsubscribeAll

### PR DESCRIPTION
**Pull request by tjmehta**
useful for using ponos in api side-by-side w/ existing workers

### CODE REVIEW

File: test/unit/server.js [hunk 1]
```diff
@@ -211,6 +211,65 @@ describe('Server', function () {
...
+
+    beforeEach(function () {
+      // runnable-hermes 6.1.0 introduced .getQueues
+      sinon.stub(hermes, 'hermesSingletonFactory').returns({
+        unsubscribe: noop,
+        getQueues: sinon.stub().returns(queues)
+      });
+      server = new ponos.Server({ queues: queues });
+      server.setAllTasks({ a: noop });
+      sinon.stub(server.hermes, 'unsubscribe');
```
Comment (bkendall) [inline test/unit/server.js]: replace `unsubscribe: noop` w/ `unsubscribe: sinon.stub()` above. then I don't think you need the `server.hermes.unsubscribe.restore()` below, either...

File: test/unit/server.js [hunk 2]
```diff
@@ -211,6 +211,65 @@ describe('Server', function () {
...
+
+    it('should call `_unsubscribe` for each queue', function (done) {
+      server._unsubscribeAll()
+        .then(function () {
+          assert.ok(server._unsubscribe.calledTwice);
+          assert.ok(server._unsubscribe.calledWith('a'));
+          assert.ok(server._unsubscribe.calledWith('b'));
+          done();
+        })
+        .catch(done);
```
Comment (bkendall) [inline test/unit/server.js]: `.catch(done)` is only something you have to do for `lab`. this is `mocha` with `chai-as-promised`!

```javascript
assert.isFulfilled(server._unsubscribeAll())
  .then(function () {
    // ...
  });
```

if it throws, you will see it in the test output. I'll double check that, but I'm fairly sure...

**Comment by bkendall** [comment]
Okay, I went and just made a few small changes. One last comment I have: you add this `_unsubscribe` method as a 'private' method, but never actually use it. Since it's undocumented (and 'private'), it's not actually that useful, yet. Would this be something we should do in the `stop()` handler? Before we close the connection to rabbitmq?

**Comment by tjmehta** [comment]
wdym?

**Comment by tjmehta** [comment]
it's documented and used by _unsubscribeAll

**Comment by tjmehta** [comment]
oh

**Comment by tjmehta** [comment]
using _unsubscribeAll in stop now.

File: lib/server.js [hunk 3]
```diff
@@ -119,6 +119,29 @@ Server.prototype._subscribeAll = function () {
...
+ * Helper function that unsubscribes to a given queue name.
+ * @param {string} queueName Name of the queue.
+ * @return {promise} A promise that resolves post worker creation
+ *   and subscription
+ */
+Server.prototype._unsubscribe = function (queueName) {
+  debug('_unsubscribe(' + queueName + ')');
+  this.log.trace('ponos.Server: unsubscribing from ' + queueName);
+  var handler = null; // all
+  return this.hermes.unsubscribeAsync(queueName, handler);
```
Comment (bkendall) [inline lib/server.js]: > `// all`

what does that mean?

File: test/unit/server.js [hunk 4]
```diff
@@ -211,6 +211,64 @@ describe('Server', function () {
...
+
+    beforeEach(function () {
+      // runnable-hermes 6.1.0 introduced .getQueues
+      sinon.stub(hermes, 'hermesSingletonFactory').returns({
+        unsubscribe: sinon.stub(),
+        getQueues: sinon.stub().returns(queues)
+      });
+      server = new ponos.Server({ queues: queues });
+      server.setAllTasks({ a: noop });
+      sinon.stub(server, '_runWorker');
```
Comment (bkendall) [inline test/unit/server.js]: `_runWorker` is not called by subscribing. only when a job is queued. You probably can safely remove this stub and the check below

File: test/unit/server.js [hunk 5]
```diff
@@ -211,6 +211,64 @@ describe('Server', function () {
...
+      });
+      server = new ponos.Server({ queues: queues });
+      sinon.stub(server, '_unsubscribe');
+      return server.setAllTasks({ a: noop, b: noop });
+    });
+
+    afterEach(function () { hermes.hermesSingletonFactory.restore(); });
+
+    it('should call `_unsubscribe` for each queue', function (done) {
+      assert.isFulfilled(
```
Comment (bkendall) [inline test/unit/server.js]: `return assert.isFulfilled`

Comment (bkendall) [inline test/unit/server.js]: also, `assert.isFulfilled()` returns a promise (surprise surprise) that you can chain off of. So your test would look like:

```javascript
return assert.isFulfilled(server._unsubscribeAll())
  .then(function () {
    /* various assertions */
  })
```

File: test/unit/server.js [hunk 6]
```diff
@@ -211,6 +211,64 @@ describe('Server', function () {
...
+        server._unsubscribeAll()
+          .then(function () {
+            assert.ok(server._unsubscribe.calledTwice);
+            assert.ok(server._unsubscribe.calledWith('a'));
+            assert.ok(server._unsubscribe.calledWith('b'));
+            done();
```
Comment (bkendall) [inline test/unit/server.js]: and then, remove `done` :)

**Comment by tjmehta** [comment]
@bkendall fixed your comments

File: test/unit/server.js [hunk 7]
```diff
@@ -211,6 +211,61 @@ describe('Server', function () {
...
+      server = new ponos.Server({ queues: queues });
+      server.setAllTasks({ a: noop });
+    });
+
+    afterEach(function () {
+      hermes.hermesSingletonFactory.restore();
+    });
+
+    it('should apply the correct callback for the given queue', function () {
+      server._unsubscribe('a');
```
Comment (bkendall) [inline test/unit/server.js]: this returns a promise, yes? you should assert on it :) otherwise, it could throw and cause a weird error in the tests

**Comment by tjmehta** [comment]
@bkendall approves

**Pull request merged**